internal/cmd/envoy/shutdown_manager.go

```diff
@@ -0,0 +1,234 @@
+
+	// Since we cannot establish order between preStop hooks, wait a bit giving the exec
+	// preStop hook a chance to touch the signaling file before starting to poll for it.
+	time.Sleep(2 * time.Second)
+
+	// Exit early if active file failed to show up
+	if _, err := os.Stat(activeFile); os.IsNotExist(err) {
+		logger.Info("shutdown active file not found")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	// Poll for shutdown readiness
+	for {
+		// Check if ready timeout is exceeded
+		elapsedTime := time.Since(startTime)
+		if elapsedTime > readyTimeout {
+			logger.Info("shutdown readiness timeout exceeded")
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+
+		_, err := os.Stat(activeFile)
+		switch {
+		case os.IsNotExist(err):
+			logger.Info("shutdown readiness detected")
+			return
+		case err != nil:
+			logger.Error(err, "error checking for shutdown readiness")
+		case err == nil:
+			time.Sleep(1 * time.Second)
+		}
+	}
+}
+
+// Shutdown is called from a preStop hook on the shutdown-manager container where
+// it will initiate a graceful drain sequence on the Envoy proxy and block until
+// connections are drained or a timeout is exceeded.
+func Shutdown(drainTimeout time.Duration, minDrainDuration time.Duration, exitAtConnections int) error {
+	var startTime = time.Now()
+	var allowedToExit = false
+
+	// Reconfigure logger to write to stdout of main process if running in Kubernetes
+	if _, k8s := os.LookupEnv("KUBERNETES_SERVICE_HOST"); k8s && os.Getpid() != 1 {
+		logger = logging.FileLogger("/proc/1/fd/1", "shutdown-manager", v1alpha1.LogLevelInfo)
+	}
+
+	// Signal to shutdownReadyHandler that drain process is started
+	if _, err := os.Create(ShutdownActiveFile); err != nil {
+		logger.Error(err, "error creating shutdown active file")
+		return err
+	}
+
+	logger.Info(fmt.Sprintf("initiating graceful drain with %.0f second minimum drain period and %.0f second timeout",
+		minDrainDuration.Seconds(), drainTimeout.Seconds()))
+
+	// Send request to Envoy admin API to initiate the graceful drain sequence
+	if resp, err := http.Post(fmt.Sprintf("http://%s:%d/drain_listeners?graceful&skip_exit",
+		bootstrap.EnvoyAdminAddress, bootstrap.EnvoyAdminPort), "application/json", nil); err != nil {
+		logger.Error(err, fmt.Sprintf("error %s", "initiating graceful drain"))
+	} else {
+		if resp.StatusCode != http.StatusOK {
+			logger.Error(fmt.Errorf("unexpected response status: %s", resp.Status), fmt.Sprintf("error %s", "initiating graceful drain"))
+		}
+		resp.Body.Close()
+	}
+
+	// Poll total connections from Envoy admin API until minimum drain period has
+	// been reached and total connections reaches threshold or timeout is exceeded
+	for {
+		elapsedTime := time.Since(startTime)
+
+		conn, err := getTotalConnections()
+		if err != nil {
+			logger.Error(err, "error getting total connections")
+		}
+
+		if elapsedTime > minDrainDuration && !allowedToExit {
```

Comment:
Probably a matter of personal preference, but I opted for this method so that drain logs make it obvious what is happening and one gets visibility on connection counts changing more immediately after starting the graceful drain vs waiting X seconds silently before starting to log for the drain process.